packages/opentelemetry-plugin-mysql/src/mysql.ts

```diff
@@ -0,0 +1,266 @@
+          pool,
+          'getConnection',
+          thisPlugin._patchGetConnection(pool)
+        );
+
+        return pool;
+      };
+    };
+  }
+
+  // global export function
+  private _patchCreatePoolCluster() {
+    return (originalCreatePoolCluster: typeof mysqlTypes.createPoolCluster) => {
+      const thisPlugin = this;
+      thisPlugin._logger.debug(
+        'MysqlPlugin#patch: patched mysql createPoolCluster'
+      );
+      return function createPool(_config: string | mysqlTypes.PoolConfig) {
+        //@ts-ignore
+        const cluster = originalCreatePoolCluster(...arguments);
+
+        shimmer.wrap(
+          cluster,
+          'getConnection',
+          thisPlugin._patchGetConnection(cluster)
+        );
+
+        return cluster;
+      };
+    };
+  }
+
+  // method on cluster or pool
+  private _patchGetConnection(pool: mysqlTypes.Pool | mysqlTypes.PoolCluster) {
+    return (originalGetConnection: Function) => {
+      const thisPlugin = this;
+      thisPlugin._logger.debug(
+        'MysqlPlugin#patch: patched mysql pool getConnection'
+      );
+      return function getConnection(
+        arg1?: unknown,
+        arg2?: unknown,
+        arg3?: unknown
+      ) {
+        if (arguments.length === 1 && typeof arg1 === 'function') {
+          const patchFn = thisPlugin._getConnectionCallbackPatchFn(arg1);
+          return originalGetConnection.call(pool, patchFn);
+        }
+        if (arguments.length === 2 && typeof arg2 === 'function') {
+          const patchFn = thisPlugin._getConnectionCallbackPatchFn(arg2);
+          return originalGetConnection.call(pool, arg1, patchFn);
+        }
+        if (arguments.length === 3 && typeof arg3 === 'function') {
+          const patchFn = thisPlugin._getConnectionCallbackPatchFn(arg3);
+          return originalGetConnection.call(pool, arg1, arg2, patchFn);
+        }
+
+        return originalGetConnection.apply(pool, arguments);
+      };
+    };
+  }
+
+  private _getConnectionCallbackPatchFn(cb: Function) {
+    const thisPlugin = this;
+    return function() {
+      if (arguments[1]) {
+        shimmer.wrap(
+          arguments[1],
+          'query',
+          thisPlugin._patchQuery(arguments[1])
+        );
+      }
+      if (typeof cb === 'function') {
+        cb(...arguments);
+      }
+    };
+  }
+
+  private _patchQuery(connection: mysqlTypes.Connection | mysqlTypes.Pool) {
+    return (originalQuery: Function): mysqlTypes.QueryFunction => {
+      const thisPlugin = this;
+      thisPlugin._logger.debug('MysqlPlugin: patched mysql query');
+
+      return function query(
+        query: string | mysqlTypes.Query | mysqlTypes.QueryOptions,
+        _valuesOrCallback?: unknown[] | mysqlTypes.queryCallback,
+        _callback?: mysqlTypes.queryCallback
+      ) {
+        const spanName = getSpanName(query);
+
+        const span = thisPlugin._tracer.startSpan(spanName, {
+          parent: thisPlugin._tracer.getCurrentSpan() || undefined,
```

Comment:
side issue: Since this pattern is repeated a lot, I wonder if we should create an issue to allow parent to be `null` as well. This way we can avoid always appending `|| undefined`?

Reply:
@markwolff Yeah i think it should be easier since every plugin has that `|| undefined`

Reply:
#565